src/cargo/core/source/source_id.rs

```diff
@@ -680,7 +680,13 @@ impl<'a> fmt::Display for SourceIdAsUrl<'a> {
                 kind: SourceKind::Registry,
                 ref url,
                 ..
-            } => write!(f, "registry+{}", url),
+            } => {
+                if url.scheme().starts_with("sparse+") {
+                    write!(f, "{}", url)
+                } else {
+                    write!(f, "registry+{}", url)
+                }
```

Comment:
(nit)
Looks reasonable but a bit subtle. Should we document the output URL as comments? Such as

```suggestion
                // For sparse http registry the URL already contains the prefix.
                if url.scheme().starts_with("sparse+") {
                    // e.g. sparse+http://example.com
                    write!(f, "{url}")
                } else {
                    // e.g. registry+http://example.com
                    write!(f, "registry+{url}")
                }
```

Reply:
Yes, thanks!